qualtran/bloqs/generalized_qsp.py

```diff
@@ -220,23 +215,21 @@ def qsp_phase_factors(
     phi = np.zeros(n)
     lambd = 0
 
+    def safe_angle(x):
+        return 0 if np.isclose(x, 0) else np.angle(x)
+
     for d in reversed(range(n)):
         assert S.shape == (2, d + 1)
 
         a, b = S[:, d]
         theta[d] = np.arctan2(np.abs(b), np.abs(a))
         # \phi_d = arg(a / b)
-        # TODO this is numerically unstable for values of a and b around 0,
-        #      np.angle(0) = 0, but for near-zero values it produces the "correct angle",
-        #      even though we want it to be 0.
-        #      There should be a more numerically stable way to compute the relevant phis
-        #      so that the final QSP sequence is valid.
-        phi[d] = np.angle(a) - np.angle(b)
+        phi[d] = 0 if np.isclose(np.abs(b), 0) else safe_angle(a) - safe_angle(b)
```

Comment:
`np.angle(x)` always returns values in `(-pi, pi]` but this would not be true for `angle(a) - angle(b)`. I guess this is not an issue for us because we use the angles for single qubit rotations and they take care of 2 * pi difference by default. 

Do you know if subtracting the angles is strictly better than directly using `np.angle(a * np.conj(b) / np.abs(b))` ?

Reply:
Not sure if it is strictly better, I just wanted to avoid multiplication/division to reduce precision errors. But I think `np.angle(a * np.conj(b))` should also work equally well (can drop the `abs` as it does not affect angle).

Reply:
I'd vote to change it to more closely match the algorithm description in the paper and keep all the phi's in `(-pi, pi]`

Reply:
updated

Reply:
actually even `angle(a / b)` works (on tests), I can update it to that as well if it's more preferred.

Reply:
I think this is fine